verovio: init at 4.2.0

tshaynik opened the pull request:
Add verovio tool for engraving MEI music notation as SVG

###### Description of changes


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Add a package for the [verovio](https://www.verovio.org) command line tool.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tshaynik committed:
verovio: init at 4.2.0

Add verovio tool for engraving MEI music notation as SVG

drupol commented on pkgs/tools/audio/verovio/default.nix:
```diff
@@ -0,0 +1,55 @@
+{ stdenv, lib, fetchFromGitHub, cmake, git, makeWrapper }:
+
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

drupol commented on pkgs/tools/audio/verovio/default.nix:
```diff
@@ -0,0 +1,55 @@
+    runHook postBuild
+  '';
+
+  installPhase = ''
```
Missing `preInstall` hook.

drupol commented on pkgs/top-level/all-packages.nix:
```diff
@@ -37203,6 +37203,8 @@ with pkgs;
 
   vhs = callPackage ../applications/misc/vhs { };
 
+  verovio = callPackage ../tools/audio/verovio { };
```
Please move it to the right place, `verovio` comes before `vhs`.

drupol commented:
Hi,

I set the PR to draft until all the comments are addressed.

Feel free to reach out if you need some help or assistance.

tshaynik commented:
Hi @drupol, just found this old PR, bumped the version of Verovio to the latest version, switched it to the new by-name convention, and addressed your concerns.

drupol reviewed (changes requested):
One last thing...

Could you pass the new nix files only through nixfmt-rfc-style?

tshaynik commented:
Formatted! Thanks a lot!

drupol commented on pkgs/by-name/ve/verovio/package.nix:
```diff
@@ -0,0 +1,68 @@
+
+  src = fetchFromGitHub {
+    owner = "rism-digital";
+    repo = finalAttrs.pname;
```
Please don't use this.

`repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

drupol commented on pkgs/by-name/ve/verovio/package.nix:
```diff
@@ -0,0 +1,68 @@
+
+  buildPhase = ''
+    runHook preBuild
+    set -exo pipefail
```
I don't think this is necessary

tshaynik replied:
yes, I think this was left over from testing. Removing now.

drupol commented on pkgs/by-name/ve/verovio/package.nix:
```diff
@@ -0,0 +1,67 @@
+
+  dontConfigure = true;
+
+  buildPhase = ''
+    runHook preBuild
+
+    cd tools
+    patchShebangs .
+    cmake ../cmake
+    make -j 8
+
+    runHook postBuild
+  '';
```
I'm not sure about this part. I'm pretty sure that this can be simplified.

tshaynik replied:
This is directly from the [instructions for building from source in the Verovio docs](https://book.verovio.org/installing-or-building-from-sources/command-line.html), with an extra `patchShebangs` thrown in to patch the tools.

I just confirmed that the build still fails without that `patchShebangs`.

FliegendeWurst commented on pkgs/by-name/ve/verovio/package.nix:
```diff
@@ -0,0 +1,67 @@
+  enableParallelBuilding = true;
+
+  meta = with lib; {
+    description = ''
+      A fast, portable and lightweight library for engraving Music Encoding Initiative (MEI) digital scores into SVG images.
+    '';
```
```suggestion
    description = "Fast, portable and lightweight library for engraving Music Encoding Initiative (MEI) digital scores into SVG images";
```

FliegendeWurst commented on pkgs/by-name/ve/verovio/package.nix:
```diff
@@ -0,0 +1,67 @@
+    owner = "rism-digital";
+    repo = "verovio";
+    rev = "version-${finalAttrs.version}";
+    hash = "sha256-PjBtnt46pHV3rDlwltVxfUMlgNoZGnI9CfnFtk2/j3s=";
```
```suggestion
    hash = "sha256-x9YcDnYG4S0kwtXwXJo223yEIqSZPoZCbbpsM5Po04M=";
```
If possible, avoid leaveDotGit = true. Perhaps the git use in CMake can be patched out.
